### PR TITLE
Final(?) tweaks for processing legacy (CSV) stories for 2023

### PR DESCRIPTION
**Pull request by philbudne**
NOTE!  Adds RSSEntry fields to preserve feed/source id data in CSV files.
PLEASE comment on the question at https://github.com/philbudne/story-indexer/blob/hist-update/indexer/story.py#L119 (file_name field)
Fetcher will also accept feed_url, feed_id, source_id from RSS file `<source>` tag.
Added --rss-file option to fetcher, to test parsing new RSS files not yet in production.
Set RSSEntry.fetch_date to the date in the name of the CSV file.
Cleaned up version picking (not needed for 2023) to use date from CSV file column instead of fetch_date.
hist-queuer: let hist-fetcher validate the downloads_id
hist-fetcher: quarantine stories with invalid downloads_id
parser: drop messages with no HTML (count only, no quarantine)

### CODE REVIEW

**Comment by pgulley** [comment]
Maybe `system_source_name` or `mc_source_name` plus a comment, as it's carrying the source of that record within the larger mc system?

File: indexer/workers/hist-queuer.py [hunk 1]
```diff
@@ -29,26 +33,46 @@ def process_file(self, fname: str, fobj: BinaryIO) -> None:
...
-                logger.error("bad downloads_id: %r", row)
-                self.incr_stories("bad-dlid", url)
-                continue
+            # let hist-fetcher quarantine if bad
+
+            try:
+                feeds_id = int(row["feeds_id"])
+            except (KeyError, ValueError):
+                # XXX cannot count w/ incr_stories (only incremented once per story)
+                feeds_id = None
```
Comment (kilemensi) [inline indexer/workers/hist-queuer.py:59]:
```python
feeds_id = int(row["feeds_id"]) if row.get("feeds_id") else None
```
Without try/except

File: indexer/workers/hist-queuer.py [hunk 2]
```diff
@@ -29,26 +33,46 @@ def process_file(self, fname: str, fobj: BinaryIO) -> None:
...
+        else:
+            fetch_date = None
+
         # typical columns:
         # collect_date,stories_id,media_id,downloads_id,feeds_id,[language,]url
         for row in csv.DictReader(io.TextIOWrapper(fobj)):
             logger.debug("%r", row)
 
-            url = row.get("url", "")
+            url = row.get("url", None)
```
Comment (kilemensi) [inline indexer/workers/hist-queuer.py:48]:
```suggestion
            url = row.get("url")
```

Isn't `None` the default?

**Review by kilemensi — APPROVED** [review]

File: indexer/story.py [hunk 3]
```diff
@@ -109,7 +109,17 @@ class RSSEntry(StoryData):
...
+
+    # none of the following are imported/indexed as of 2/2024;
+    # available for internal tracing/debugging only.
+    source_url: Optional[str] = None  # source tag url property
+    source_feed_id: Optional[int] = None  # source tag mcFeedId property
+    source_source_id: Optional[int] = None  # source tag mcFeedId property
+    # NOTE! If we ever move to rss-fetcher delivering stories by queue message,
+    # there won't be a "file".  "source_name" might be a better field name,
+    # but could easily be confused as comming from the RSS <source> tag. IDEAS???
+    file_name: Optional[str] = None  # full name or URL of source (RSS or CSV) file
```
[the remark this answers is not in the document]
Comment (philbudne) [reply indexer/story.py:122]: I like it!  Also: `from`, and `via` (which carries the idea that it isn't the ORIGINAL source of the info, just the conduit)

File: indexer/workers/hist-queuer.py (hunk 1 above)
Comment (philbudne) [reply indexer/workers/hist-queuer.py:59]: 1. I avoid the Python trinary (I find the condition in the middle hard to read)
2. I wanted to catch the possibility the a value is supplied, but not a valid number
3. I thought I had a third point. Maybe I'll remember it after I finish my coffee?!

File: indexer/story.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (pgulley) [reply indexer/story.py:122]: iterating- `loaded_from` ?

**Review by pgulley — APPROVED** [review]

File: indexer/story.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (philbudne) [reply indexer/story.py]: as mentioned in meeting, decided on via (used as a header name in the Session Initiation Protocol used in VoIP).

**Comment by philbudne** [comment]
@thepsalmist any review comments?

**Comment by thepsalmist** [comment]
> @thepsalmist any review comments?

LGTM!

**Review by thepsalmist — APPROVED** [review]

**Pull request merged**